pkg/operator/credentials_test.go

```diff
@@ -0,0 +1,321 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package operator
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	cco "github.com/openshift/cloud-credential-operator/pkg/apis/cloudcredential/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	"github.com/openshift/aws-load-balancer-operator/pkg/utils/test"
+)
+
+func TestProvisionCredentials(t *testing.T) {
+	tests := []struct {
+		name    string
+		envVars map[string]string
+		scheme  *runtime.Scheme
+		// provisionedSecret simulates CCO job
+		// which is supposed to create a secret from CredentialsRequest
+		provisionedSecret   *corev1.Secret
+		expectedCredReqName types.NamespacedName
+		compareCredReq      func(*cco.CredentialsRequest, *cco.AWSProviderSpec) error
+		expectedContents    string
+		errExpected         bool
+	}{
+		{
+			name: "nominal sts",
+			envVars: map[string]string{
+				"ROLEARN": "arn:aws:iam::123456789012:role/foo",
+			},
+			scheme: test.Scheme,
+			provisionedSecret: &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "aws-load-balancer-operator",
+					Namespace: "aws-load-balancer-operator",
+				},
+				Data: map[string][]byte{
+					"credentials": []byte("oksts"),
+				},
+			},
+			expectedCredReqName: types.NamespacedName{
+				Namespace: "openshift-cloud-credential-operator",
+				Name:      "aws-load-balancer-operator",
+			},
+			compareCredReq: func(credReq *cco.CredentialsRequest, providerSpec *cco.AWSProviderSpec) error {
```

Comment:
Indeed why this way? It's so clever that even I forgot why I did so :) That's a good sign that this function deserves a comment.

I think my idea was based on the fact that I don't like to test things which don't change. We create a desired `CredentialsRequest` in `credentials.go`, there is only one copy of it so we either made a mistake in it or we don't. Repeating the same `CredentialsRequest` definition in `credentials_test.go` won't test anything - if the error is there we'll just repeat it in tests. The desired `CredentialsRequest` should be  (and it is) tested in the e2e test. So here I just tested the moving part of the desired `CredentialsRequest`. This approach helps to avoid the unnecessary repetition of the desired definition in each test case. But I agree it looks less intuitive.